### DISCUSSION TITLE
CBR to Pdf conversion failed for "cache resources exhausted"

### DISCUSSION THREAD

**Discussion by rabbygit** [ImageMagick/ImageMagick#6085] [Help]
### ImageMagick version

7.1.0-62

### Operating system, version and so on

20.04.3 LTS

### Description

I’m trying to convert a `275 MB` CBR file to pdf but the conversion is getting failed. These are the steps I followed:

1. Command to unzip the rar: unrar e '/tmp/1676456470685.2727/6782.rar' /tmp/1676456470889.922
2. Command to convert the unzipped jpg images to pdf:
```
convert /tmp/1676456470889.922/*.jpg -profile /app/sRGB2014.icc -strip -set profile /app/sRGB2014.icc -compress jpeg -resize 1240.2x1753.95 
-gravity Center -extent 1240.2x1753.95 -units PixelsPerInch -density 150 -repage 1240.2x1753.95 -bordercolor white -border 0 output.pdf
```
Error:
```
convert: cache resources exhausted `/tmp/1676456470889.922/OPSX-UK-028_Page_154.jpg' @ error/cache.c/OpenPixelCache/3879.
convert: cache resources exhausted `/tmp/1676456470889.922/OPSX-UK-028_Page_155.jpg' @ error/cache.c/OpenPixelCache/3879.
convert: cache resources exhausted `/tmp/1676456470889.922/OPSX-UK-028_Page_156.jpg' @ error/cache.c/OpenPixelCache/3879.
convert: cache resources exhausted `/tmp/1676456470889.922/OPSX-UK-028_Page_001.jpg' @ error/cache.c/OpenPixelCache/3879.
```
### About system:
Host machine:
OS: Ubuntu 22.04.1 LTS
Memory: 64GB

ImageMagick is running in a docker container:
Container OS: Ubuntu 20.04.3 LTS
ImageMagick version: 7.1.0-62

Current ImageMagick resource policy:
```
<policy domain="resource" name="memory" value="16GiB"/>
<policy domain="resource" name="map" value="32GiB"/>
<policy domain="resource" name="area" value="24GB"/>
<policy domain="resource" name="disk" value="64GiB"/>
```

Even when we increase the resource limits further, the same error happens. Is there a way that we can optimize our command so that we can avoid resource exhaustion?

**Comment by snibgo** [comment]
You have IM v7, but use "convert", which uses the v6 syntax. I suggest using "magick" instead, as this gives you the more powerful and logical v7 syntax.

"cache resources exhausted" means your computer doesn't have enough space (memory and disk) for all the pixels you are processing.

How many pixels are in `/tmp/1676456470889.922/*.jpg`? How many files, and how many pixels in a typical file?

You are trying to read all those images into memory at the same time. One alternative is to use "`magick mogrify`" to process the files one at a time. This can create one PDF for each JPG. Then use pdfunite to read those PDFs and create one final PDF.

**Reply by rabbygit** [reply]
> You have IM v7, but use "convert", which uses the v6 syntax. I suggest using "magick" instead, as this gives you the more powerful and logical v7 syntax.
> 
> "cache resources exhausted" means your computer doesn't have enough space (memory and disk) for all the pixels you are processing.
> 
> How many pixels are in `/tmp/1676456470889.922/*.jpg`? How many files, and how many pixels in a typical file?
> 
> You are trying to read all those images into memory at the same time. One alternative is to use "`magick mogrify`" to process the files one at a time. This can create one PDF for each JPG. Then use pdfunite to read those PDFs and create one final PDF.

Hey, Thank you for your response. 
On this directory `/tmp/1676456470889.922`, there are 158 jpg files and each has an average 3578px * 4377px.

I've tried your suggested way with this command, but it also failed after converting 99 files to 99 pdf and then throwing `unable to extend cache '/imgs/OPSX-UK-028_Page_004.jpg': No space left on device @ error/cache.c/OpenPixelCache/3910`. It seems like the image is not processed one at a time as you said. 

```
magick mogrify /tmp/1676456470889.922/*.jpg -profile /app/sRGB2014.icc -strip -set profile /app/sRGB2014.icc -compress jpeg -resize 1240.2x1753.95 
-gravity Center -extent 1240.2x1753.95 -units PixelsPerInch -density 150 -repage 1240.2x1753.95 -bordercolor white -border 0 output.pdf
```

**Comment by snibgo** [comment]
158 images each  3578px * 4377px is about 2 billion pixels. You might need 12 bytes/pixel, so that would be 24 GB of memory. "`magick -list resource`" will tell you the limits of memory and disk that IM can use, if it is available.

Your `magick mogrify` command is bad. It should end with the input files, which might be a list or wildcard filename, with path if needed. It does not take an output filename, because it will name each output after the corresponding input. See https://imagemagick.org/script/mogrify.php

**Comment by urban-warrior** [comment]
See https://imagemagick.org/script/architecture.php#tera-pixel, a discussion on how to deal with large images and image sequences.
